src/quorumnet/sn_network.cpp

```diff
@@ -0,0 +1,1019 @@
+                    // ["", "ROUTE", "CMD"] -- for an incoming message from a non-SN node (i.e. not a SN quorum message)
+                    //
+                    // The latter two may be optionally followed by one frame containing a serialized bt_dict.
+                    //
+                    // The pubkey form supports sending a reply back to the given PUBKEY even if the
+                    // original connection is closed -- a new connection to that SN will be
+                    // established if required.  The routed form only supports replying on the
+                    // existing incoming connection (any attempted reply will be dropped if the
+                    // connection no longer exists).
+
+                    std::string pubkey;
+                    if (i == 1) { // Talking to ourself
+                        pubkey = this->pubkey;
+                    } else {
+                        try {
+                            const char *pubkey_hex = parts.back().gets("User-Id");
+                            auto len = std::strlen(pubkey_hex);
+                            assert(len == 0 || len == 64);
+                            if (len == 64)
+                                pubkey = from_hex(pubkey_hex, pubkey_hex + 64);
+                        } catch (...) {} // User-Id not set, i.e. no pubkey
+                    }
+
+                    if (!pubkey.empty()) {
+                        // SN message which means we want to stick the pubkey on the front.  For a
+                        // connection on the listener we also want to drop the first part (the
+                        // return route) because SN replies have stronger routing.
+                        if (i <= 1) // listener or self: discard the return route
+                            parts.pop_front();
+                        parts.emplace_front(pubkey.data(), pubkey.size());
+
+                        if (parts.size() < 2 || parts.size() > 3) {
+                            SN_LOG(warn, "Invalid incoming message; expected 1-2 parts, not " << (parts.size() - 1));
+                            continue;
+                        }
+
+                        auto it = peers.find(pubkey);
+                        if (it != peers.end())
+                            it->second.activity();
+
+                    } else {
+                        // No pubkey (i.e. not a SN quorum message); this can only happen on an
+                        // incoming connection on listener, ...
+                        assert(i == 0);
+                        // ... which means the first part is already the return route, so just leave
+                        // it there and prepend a blank frame to indicate a no-pubkey message.
+                        parts.emplace_front();
+
+                        if (parts.size() < 3 || parts.size() > 4) {
+                            SN_LOG(warn, "Invalid incoming message: expected 1-2 parts, not " << (parts.size() - 2));
```

Comment:
Similar to above, but here zmq has already prepended an internal return routing marker and we're pushing another blank part before that to tag the message as not being from an SN source.  The actual "meat" of the message (i.e. what the other side sent) is then 2 less than the parts we have in the `parts` list.  I can't do the same change as above here, though; even if I move the empty prepending down it's still going to be off by one between the conditions and the message.

Reply:
Its confusing because of the reusing of parts whilst also doing error checking. If we could classify the message first, and then error checking on the message type, then updating inplace if necessary  then forwarding to worker separates it into distinct steps it'd be clearer.

Reply:
Agreed.  Though in this case it's ZMQ itself that is adding one of those parts (in the previous case I reordered the error/check to make more sense).  In this case I've rewritten it with a separate calculation and comment (not pushing yet because they are on top of the blink branch and stash + branch switching right now is screwing up):
```C++
auto incoming_parts = parts.size() - 2; // ZMQ router prepended one, we just prepended another.

if (incoming_parts < 1 || incoming_parts > 2) {
    SN_LOG(warn, "Invalid incoming message: expected 1-2 parts, not " << incoming_parts);
    continue;
}
```